### PR TITLE
"reliable" mqtt rebased

### PR DESCRIPTION
**Pull request by linuxbasic**

### CODE REVIEW

File: cerk_port_mqtt/README.md [hunk 1]
```diff
@@ -49,48 +55,25 @@ E.g. `Config::String(String::from("test"))`
...
 * 2: Exactly once delivery
 
 E.g. `Config::U8(0)`
 
-### subscribe_topics
+#### subscribe_topic
 
-The value has to by of type `Config::Vec([Config::String])` and must have the same length as `subscribe_qos`.
-The values in the vector contain the MQTT topic wich the router should subscribe to.
+The value has to by of type `Config::String` and contain the MQTT topic name  which the router should subscribe to.
```
Comment (Lazzaretti) [inline cerk_port_mqtt/README.md:71]:
```suggestion
The value has to be of type `Config::String` and contains the MQTT topic name to which the router should subscribe to.
```

File: cerk_port_mqtt/README.md [hunk 2]
```diff
@@ -49,48 +55,25 @@ E.g. `Config::String(String::from("test"))`
...
+#### subscribe_qos
 
 The [quality of service](http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html#_Toc398718099) for the topic subscription.
-The quality of service is only for the MQTT broker and does not change any behavior of the router or the port.
-The router only supports best effort at the moment.
-
-* 0: At most once delivery
-* 1: At least once delivery
-* 2: Exactly once delivery
+The value has to by of type `Config::U8` and currently only `0` (At most once delivery) is supported. (see https://github.com/ce-rust/cerk/issues/71)
```
Comment (Lazzaretti) [inline cerk_port_mqtt/README.md:76]:
```suggestion
The value has to be of type `Config::U8` and currently only `0` (At most once delivery) is supported. (see https://github.com/ce-rust/cerk/issues/71)
```

File: cerk_port_mqtt/README.md [hunk 3]
```diff
@@ -185,11 +153,11 @@ let config = Config::HashMap(map);
 
 ## Examples
 
-* [Generator to MQTT](https://github.com/ce-rust/cerk/tree/master/examples/examples/src/mqtt/)
+* [Generator to MQTT](https://github.com/ce-rust/cerk/tree/master/examples/src/mqtt/)
 
 ## Limitations
 
-* **reliability** this port does not support any `DeliveryGuarantee` other then `BestEffort` and so does never send a `OutgoingCloudEventProcessed` or `IncomingCloudEventProcessed` messages
+* **reliability** this port does not support any `DeliveryGuarantee` other then `BestEffort` and so does never send a `OutgoingCloudEventProcessed` or `IncomingCloudEventProcessed` messages.
```
Comment (Lazzaretti) [inline cerk_port_mqtt/README.md:160]: before you wrote it supports sending with at least once?

File: cerk_port_mqtt/README.md [hunk 4]
```diff
@@ -49,48 +55,25 @@ E.g. `Config::String(String::from("test"))`
...
-
 #### send_qos
 
 The [quality of service](http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html#_Toc398718099) for message delivery.
 The quality of service is only for the MQTT broker and does not change any behavior of the router or the port.
-The router only supports best effort at the moment.
 
 * 0: At most once delivery (default)
 * 1: At least once delivery
 * 2: Exactly once delivery
```
Comment (Lazzaretti) [inline cerk_port_mqtt/README.md:65]: It makes no sense if the router does not support it in general

**Review by Lazzaretti — CHANGES_REQUESTED** [review]

File: cerk_port_mqtt/src/port_mqtt.rs [hunk 5]
```diff
@@ -215,13 +172,47 @@ pub fn port_mqtt_start(id: InternalServerId, inbox: BoxedReceiver, sender_to_ker
...
+                            ));
+                        }
+                        Err(err) => panic!("{} connection setup failed {:?}", id, err),
+                    }
+                } else {
+                    panic!("{} can not send message, no connection configured", id)
+                }
+            }
+            BrokerEvent::IncomingCloudEventProcessed(event_id, result) => {
+                debug!("{} message {} processed -> {}", id, event_id, result);
```
Comment (Lazzaretti) [inline cerk_port_mqtt/src/port_mqtt.rs:215]: why?

File: examples/examples/src/mqtt/sequence_to_mqtt/main.rs [hunk 6]
```diff
@@ -27,6 +27,7 @@ fn static_config_loader_start(
             Config::String("tcp://mqtt-broker:1883".to_string()),
         ),
         ("send_topic".to_string(), Config::String("test".to_string())),
+        ("send_qos".to_string(), Config::U8(1)),
```
Comment (Lazzaretti) [inline examples/examples/src/mqtt/sequence_to_mqtt/main.rs:30]: we don't have this option, have we?

Comment (linuxbasic) [inline examples/examples/src/mqtt/sequence_to_mqtt/main.rs:30]:
```suggestion
```

File: cerk_port_mqtt/README.md [hunk 7]
```diff
@@ -185,11 +134,7 @@ let config = Config::HashMap(map);
 
 ## Examples
 
-* [Generator to MQTT](https://github.com/ce-rust/cerk/tree/master/examples/examples/src/mqtt/)
-
-## Limitations
-
-* **reliability** this port does not support any `DeliveryGuarantee` other then `BestEffort` and so does never send a `OutgoingCloudEventProcessed` or `IncomingCloudEventProcessed` messages
+* [Generator to MQTT](https://github.com/ce-rust/cerk/tree/master/examples/src/mqtt/)
```
Comment (Lazzaretti) [inline cerk_port_mqtt/README.md:137]:
```suggestion
* [Generator to MQTT](https://github.com/ce-rust/cerk/tree/master/examples/examples/src/mqtt/)
```

File: cerk_port_mqtt/src/port_mqtt.rs [hunk 8]
```diff
@@ -1,210 +1,167 @@
...
-            }
-        }
+    if let Some(ref subscribe_topic) = connection.subscribe_topic {
+        debug!("{} subscribes to {} with qos 0", id, subscribe_topic,);
+
+        connection.client.subscribe(subscribe_topic, 0).await?;
     }
 
-    (cli, options)
+    return Ok(connection);
```
Comment (Lazzaretti) [inline cerk_port_mqtt/src/port_mqtt.rs:129]:
```suggestion
    Ok(connection)
```

File: examples/examples/src/mqtt/mqtt-to-printer/main.rs [hunk 9]
```diff
@@ -26,13 +26,10 @@ fn static_config_loader_start(
...
-            "subscribe_topics".to_string(),
-            Config::Vec(vec![Config::String("test".to_string())]),
-        ),
-        (
-            "subscribe_qos".to_string(),
-            Config::Vec(vec![Config::U8(0)]),
+            "subscribe_topic".to_string(),
+            Config::String("test".to_string()),
         ),
+        ("subscribe_qos".to_string(), Config::U8(1)),
```
Comment (linuxbasic) [inline examples/examples/src/mqtt/mqtt-to-printer/main.rs:32]:
```suggestion
```

**Review by Lazzaretti — APPROVED** [review]

**Review by Lazzaretti — APPROVED** [review]

**Pull request merged**